// @inventivetalent/loading-cache/dist/cjs/cache/WrappedCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WrappedCache = void 0;
const CacheBase_1 = require("./CacheBase");
const SimpleCache_1 = require("./SimpleCache");
/**
 * Wrapper around simple string KV getter/setter/deleter functions
 * Intended for e.g. creating a persistent cache using the browser's localStorage
 */
class WrappedCache extends SimpleCache_1.SimpleCache {
    constructor(options, getter, setter, deleter, allDeleter) {
        super(options);
        this.getter = getter;
        this.setter = setter;
        this.deleter = deleter;
        this.allDeleter = allDeleter;
    }
    getEntryIfPresent(key, recordStats = this.options.recordStats) {
        const s = super.getEntryIfPresent(key, recordStats);
        if (s) {
            return s;
        }
        const v = this.getter(JSON.stringify(key));
        if (v) {
            const parsed = CacheBase_1.Entry.fromJson(key, JSON.parse(v));
            this.putEntry(key, parsed);
            return parsed;
        }
        return undefined;
    }
    putEntry(key, entry) {
        super.putEntry(key, entry);
        this.setter(JSON.stringify(key), JSON.stringify(entry));
    }
    invalidateEntry(key) {
        this.deleter(JSON.stringify(key));
        return super.invalidateEntry(key);
    }
    invalidateAll() {
        super.invalidateAll();
        if (this.allDeleter) {
            this.allDeleter();
        }
    }
}
exports.WrappedCache = WrappedCache;
