// @inventivetalent/loading-cache/dist/esm/Caches.js
import { AsyncLoadingCache, LoadingCache, SimpleCache } from "./cache";
class CacheBuilder {
    constructor() {
        this.options = {};
    }
    expireAfterAccess(expireAfterAccess) {
        this.options.expireAfterAccess = expireAfterAccess;
        return this;
    }
    expireAfterWrite(expireAfterWrite) {
        this.options.expireAfterWrite = expireAfterWrite;
        return this;
    }
    deleteOnExpiration(deleteOnExpiration) {
        this.options.deleteOnExpiration = deleteOnExpiration;
        return this;
    }
    expirationInterval(expirationInterval) {
        this.options.expirationInterval = expirationInterval;
        return this;
    }
    recordStats(recordStats) {
        this.options.recordStats = recordStats;
        return this;
    }
    build(loader, multiLoader, loadingInternalCache) {
        if (typeof loader === "undefined") {
            if (loadingInternalCache) {
                return loadingInternalCache(this.options);
            }
            return new SimpleCache(this.options);
        }
        return new LoadingCache(this.options, loader, multiLoader, loadingInternalCache);
    }
    buildAsync(loader, multiLoader, loadingInternalCache) {
        return new AsyncLoadingCache(this.options, loader, multiLoader, loadingInternalCache);
    }
}
export class Caches extends CacheBuilder {
    constructor() {
        super();
    }
    static builder() {
        return new CacheBuilder();
    }
}
